File: Problem-67.py
# Leet code- Combination Sum- 39 - https://leetcode.com/problems/combination-sum/
#Time complexity - (2^N)
#space complexity 
# Approach - Backtracking- In Backtracking we process from bottom onwards, while in DFS we process from top onwards. 



class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res=[]
        curr_list=[]
        index=0
        if len(candidates)==0: return res
        return self.backtrack(candidates,target,curr_list,res,index)
    
    def backtrack(self,candidates,target,curr_list,res,index):
        
            #base case
            if target==0:
                res.append(curr_list)
            if target<0:
                curr_list.pop()
        
            #logic
            size=len(candidates)
            for i in range(size):
                if candidates[i]<=target:
                    self.backtrack(candidates[i:],target-candidates[i],curr_list+ [candidates[i]],res,index)
            return res

File: Problem67.java
Executed in Leet code -Yes
Time complexity- o(n2)
Space Complexity- 0(n2)

class Solution {
    List<List<Integer>> result; 
    int target =0;
    public List<List<Integer>> combinationSum(int[] candidates, int targ) {
        target=targ;
        result = new ArrayList<>();
        if(target==0  || candidates==null || candidates.length==0) return result;
        helper(candidates,new ArrayList<Integer>(), 0,0);
        return result;
    }
    private void helper(int[] candidates,List<Integer> tmpResult,int sum, int index){
        if(sum>target) return;
        if(sum==target){
            result.add(new ArrayList<Integer>(tmpResult));
            return;
        }
        for(int i=index;i< candidates.length;i++){
            //List<Integer> tmp= new ArrayList<Integer>(tmpResult) ; 
            tmpResult.add(candidates[i]);
            helper(candidates,tmpResult,sum+candidates[i],i);
            tmpResult.remove(tmpResult.size()-1);
        }
    }
}
